### PR TITLE
feat(SEPA Payment Order): set discounted outstanding amounts (DRC-227)

### PR DESCRIPTION
**Pull request by PatrickDEissler**
This pull request introduces a mechanism to dynamically update SEPA payment amounts in payment orders, particularly to handle scenarios like early payment discounts. It adds backend logic to calculate the correct payment amount based on the execution date and provides a frontend workflow for users to update and confirm payment amounts before submission.

**Dynamic SEPA Payment Amount Calculation and Update Workflow**

_Backend logic and API:_

* Added `get_sepa_payment_amount` method to **Purchase Invoice** and registered it as a doc event, enabling calculation of payment amounts (including discounts) based on the execution date. 
* Implemented `get_payment_amount` and `update_payment_amounts` in `SEPAPaymentOrder`, which query the reference document for updated amounts and apply them to payments, with user feedback via messages. 

_Frontend workflow:_

* Added a hidden `should_update_amounts` field to the **SEPA Payment Order** DocType and logic to show an "Update Amounts" button when amounts may need recalculation. Also, added a confirmation dialog before submission if amounts are outdated.

_Documentation:_

* Updated the `README.md` to document the new SEPA payment amount calculation hook and its use case for early payment discounts.

These changes together ensure that SEPA payment orders reflect accurate amounts, allow users to update them easily, and provide safeguards against submitting outdated payment data.

### TODO

- [x] Translation

### CODE REVIEW

**Comment by barredterra** [comment]
Remaining issue: what should we do if payment amount changes during submit?

Agreed with @PatrickDEissler:

- Fetch outstanding amounts when creating the Payment Order
- Don't fetch automatically again, offer a button instead

**Review by barredterra — COMMENTED** [review]
## Review Summary ✅

Thank you for this well-structured PR! The implementation of dynamic SEPA payment amount calculation with early payment discount support is solid and follows the established codebase patterns well.

### CI Status:
- ✅ **All checks passing**
  - Summary: SUCCESS
  - Unit Tests: SUCCESS
  - Linters: SUCCESS

### PR Description:
- ✅ **Description is accurate and comprehensive**
- Clearly explains backend logic, API methods, and frontend workflow
- Includes documentation updates in README
- One TODO noted: Translation (expected for this stage)

### What Works Well:

1. **Excellent Pattern Adherence**: The new `get_sepa_payment_amount` hook follows the exact same pattern as the existing `sepa_payment_order_status_changed` hook - registered in `doc_events`, called via `run_method()`, with appropriate function signature.

2. **Clean Separation of Concerns**: Backend calculation logic is properly separated from UI workflow, with clear communication via the `should_update_amounts` flag.

3. **User-Friendly UX**: 
   - "Update Amounts" button only shows when needed (sepa_payment_order.js:34-42)
   - Confirmation dialog prevents accidental submission with outdated amounts (sepa_payment_order.js:45-62)
   - Clear user feedback with formatted messages

4. **Precision Handling**: Proper use of `.precision("outstanding")` for rounding discounted amounts (purchase_invoice.py:174-176)

5. **Good Documentation**: README update clearly explains the hook mechanism and use case

### Areas for Improvement:

#### 1. Missing Null Check for `new_amount` (sepa_payment_order.py:183-187)

**Location**: `banking/ebics/doctype/sepa_payment_order/sepa_payment_order.py:183-187`

```python
new_amount = doc.run_method(
    "get_sepa_payment_amount",
    payment.reference_row_name,
    getdate(execution_date) if execution_date else getdate(),
)
```

**Issue**: If the reference document doesn't implement `get_sepa_payment_amount`, `run_method()` might return `None`, which would cause line 190 to fail with a TypeError when attempting `round(new_amount, precision)`.

**Suggestion**: Add null check:
```python
new_amount = doc.run_method(
    "get_sepa_payment_amount",
    payment.reference_row_name,
    getdate(execution_date) if execution_date else getdate(),
)
if new_amount is None:
    return None

precision = payment.precision("amount")
if round(new_amount, precision) != round(payment.amount, precision):
    return new_amount
```

#### 2. Type Hint Mismatch (purchase_invoice.py:160)

**Location**: `banking/custom/purchase_invoice.py:160`

```python
def get_sepa_payment_amount(
    doc: "PurchaseInvoice", method: str, payment_schedule_row_name: str, execution_date: date
) -> float:
```

**Issue**: The type hint says `execution_date: date` but the function is called from `get_payment_amount()` with `getdate(execution_date) if execution_date else getdate()`, suggesting `execution_date` could be `None`. Also, line 62 always passes `getdate()`, but a SEPA Payment Order might have `execution_date=None` initially.

**Suggestion**: Update signature to handle None explicitly:
```python
def get_sepa_payment_amount(
    doc: "PurchaseInvoice", 
    method: str, 
    payment_schedule_row_name: str, 
    execution_date: date | None = None
) -> float:
    if execution_date is None:
        execution_date = getdate()
    
    scheduled_payment = next((p for p in doc.payment_schedule if p.name == payment_schedule_row_name), None)
    # rest of logic...
```

#### 3. Missing Test Coverage

Based on codebase analysis, there's **no dedicated test file** for the new discount calculation logic. The test file `test_sepa_payment_order.py` exists but is empty.

**Suggestion**: Add tests for:
- Payment amount calculation with valid discount (execution_date <= discount_date)
- Payment amount calculation with expired discount (execution_date > discount_date)
- Payment amount calculation with no discount_date
- Payment amount calculation when payment schedule row not found
- `should_update_amounts` flag setting when amounts differ
- `update_payment_amounts()` method updating multiple rows

Example test structure:
```python
def test_get_sepa_payment_amount_with_valid_discount(self):
    # Create Purchase Invoice with payment schedule containing discount
    # execution_date within discount period
    # Assert discounted amount is returned

def test_get_sepa_payment_amount_with_expired_discount(self):
    # Create Purchase Invoice with payment schedule
    # execution_date after discount_date
    # Assert full outstanding amount is returned
```

#### 4. Minor: Typo in PR Title

PR title has "**discunted**" instead of "**discounted**"

### Questions:

1. **What happens if the user changes `execution_date` after the Payment Order is submitted?** The `on_update_after_submit` hook only checks `transmission_datetime`, not amount changes. Should there be validation to prevent execution_date changes after submission?

2. **Should the "Update Amounts" button also refresh from the backend** (re-fetch from database) in case the Purchase Invoice's payment schedule was updated in another session?

3. **Performance consideration**: The `set_should_update_amounts()` method (line 81-84) calls `get_payment_amount()` for every payment on every validation. For Payment Orders with many payments, this could be slow. Consider caching or only checking when `execution_date` changes.

### Recommendations:

**Must Fix:**
- Add null check for `new_amount` in `get_payment_amount()` (sepa_payment_order.py:183-187)

**Should Fix:**
- Add type hint for optional `execution_date` parameter in `get_sepa_payment_amount()` 
- Add test coverage for the new discount calculation logic
- Fix typo in PR title

**Consider:**
- Optimize `set_should_update_amounts()` to only re-check when execution_date changes
- Add validation to prevent execution_date changes after submission
- Consider adding a helper method to determine if discount is still valid for better code reuse

---

Overall, this is a **well-implemented feature** that follows the established patterns in the codebase. The main concerns are around edge case handling (null checks, type hints) and test coverage. The core logic is sound and the UX is thoughtful.

Great work! 🎉

File: banking/ebics/doctype/sepa_payment_order/sepa_payment_order.py [hunk 1]
```diff
@@ -67,10 +72,32 @@ def before_validate(self):
...
+		if self.has_value_changed("execution_date"):
+			self.set_should_update_amounts()
+
 	def validate(self):
 		self.validate_ibans()
 		self.validate_account_currency()
 
+	def set_should_update_amounts(self):
+		self.should_update_amounts = 0
+		if any(get_payment_amount(payment, self.execution_date) is not None for payment in self.payments):
```
Comment (PatrickDEissler) [inline banking/ebics/doctype/sepa_payment_order/sepa_payment_order.py]: This condition is mostly True.
Therefore, even if nothing should change, the checkbox is ticked and the confirm dialog appears before submit.

File: banking/ebics/doctype/sepa_payment_order/sepa_payment_order.js [hunk 2]
```diff
@@ -30,6 +30,35 @@ frappe.ui.form.on("SEPA Payment Order", {
 			// 	__("Actions")
 			// );
 		}
+
+		if (
+			frm.doc.docstatus === 0 &&
+			frm.has_perm("write") &&
+			frm.doc.should_update_amounts
```
Comment (PatrickDEissler) [inline banking/ebics/doctype/sepa_payment_order/sepa_payment_order.js]: IMO the last part of the if-condition is not needed.
I think the button can just always be shown. It is a little less to "worry" about.

File: banking/ebics/doctype/sepa_payment_order/sepa_payment_order.py (hunk 1 above)
Comment (barredterra) [reply banking/ebics/doctype/sepa_payment_order/sepa_payment_order.py]: This is only true if at least one of the amounts changed. I've improved the function name and docstring to make this clearer.

File: banking/ebics/doctype/sepa_payment_order/sepa_payment_order.js (hunk 2 above)
Comment (barredterra) [reply banking/ebics/doctype/sepa_payment_order/sepa_payment_order.js]: I agree as a developer, but as a user I would like to see relevant buttons only.

**Comment by PatrickDEissler** [comment]
Tbd via call:

- I think, I can replicate that the checbox is mostly True.
- I also think, that if a **SEPA Payment Order** is left as draft for a few days, the button will not be shown – even if the outstanding amount should have changed. Therefore, I prefer to always show the button.

**Comment by barredterra** [comment]
Decision:

- Always show "Update Payments" button
- Remove _Should Update Payments_ checkbox
- `before_submit` checks only for changed payments

File: banking/custom/purchase_invoice.py [hunk 3]
```diff
@@ -149,3 +154,23 @@ def sepa_payment_order_status_changed(
...
+
+def get_sepa_payment_amount(
+	doc: "PurchaseInvoice", method: str, payment_schedule_row_name: str, execution_date: date
+) -> float:
+	scheduled_payment = next((p for p in doc.payment_schedule if p.name == payment_schedule_row_name), None)
+	if not scheduled_payment:
+		# No Payment Schedule to check.
+		return 0
+
+	if not scheduled_payment.discount_date or getdate(scheduled_payment.discount_date) < execution_date:
```
Comment (PatrickDEissler) [inline banking/custom/purchase_invoice.py:167]:
```suggestion
	if not scheduled_payment.discount_date or getdate(scheduled_payment.discount_date) <= execution_date:
```

**Comment by PatrickDEissler** [comment]
@barredterra 
Thanks! I have tested it. Works well.
I just suggest to adjust the condition that compares the `discount_date`.

File: banking/custom/purchase_invoice.py (hunk 3 above)
Comment (barredterra) [reply banking/custom/purchase_invoice.py:167]: @PatrickDEissler if the invoice says "2% discount for payments made by October 29", I still get the discount when I pay on October 29, right?
